### DISCUSSION TITLE
Increase of 'UNREGISTERED' error responses in web-push notificatioins since March 6th

### DISCUSSION THREAD

**Discussion by ramilyayus** [firebase/firebase-js-sdk#8099] [Q&A] [answered]
Hi there,

We're in the B2B sector, and our clients use our services to send web pushes to their audiences via FCM. We have a substantial client base numbering in the hundreds. However, for the past three weeks, we've been grappling with a Firebase issue that's significantly affecting our operations.

I'm reaching out to see if anyone else has encountered a similar issue or if you have any insights/advices on what to check or analyze in this situation. We're running out of ideas, and FCM support hasn't been helpful.

**Here's what's happening:**

- Starting March 6th, we've noticed an uptick in UNREGISTERED errors when sending web pushes.
- The frequency of other errors remains largely unchanged.
- This issue is exclusive to web pushes; our mobile pushes via FCM are unaffected.
- We haven't found any link between this error and specific browser versions, devices, or manufacturers.
- We didn't implement any code changes at that time. However, a month prior, we did switch to the Firebase API v1 for sending web pushes. We don't believe this should cause issues since it's been a while, and some clients on the older API version are also experiencing more errors.
- We're also seeing a spike in token refreshes, suggesting Firebase might be changing tokens more often, or clients are tweaking permissions off-site.
- **We have confirmation that at least one other company has experienced this exact problem.**

<details><summary>Here's an example of an FCM response we're getting:</summary>
<p>

```
{
  "code": 404,
  "message": "Requested entity was not found.",
  "status": "NOT_FOUND",
  "details": [
    {
      "@type": "type.googleapis.com/google.firebase.fcm.v1.FcmError",
      "errorCode": "UNREGISTERED"
    }
  ]
}

```

</p>
</details>

**We've also noticed two Firebase-related errors on the frontend service-worker:**

- `An issue with updating the user from FCM: "The caller does not have permission" (messaging/token-update-failed).`
- `An issue with subscribing the user to FCM: "Requested entity was not found" (messaging/token-subscribe-failed).`

We're struggling to determine the cause of these errors. Since we weren't tracking these errors before the issue arose, we can't confidently say if they're related to the increase in UNREGISTERED errors.

According to the response from Firebase support, based on several tokens, we understood that the UNREGISTERED response for them was valid because these tokens were deleted by the service worker and such behavior is expected. However, the increased number of such cases does not seem healthy to us; our only working hypothesis regarding what is happening looks like this:

- The user visits the site.
- A service worker is installed and Firebase issues a token.
- We send a push.
- The push is displayed.
- Something crashes at this moment in our code or in the Firebase JS SDK code.
- The service worker transitions to an unregistered state and signals FCM to delete the token.
- We are left with the last known token and upon trying to send a push again, we get UNREGISTERED.

**Firebase support advised us to update our Firebase JS SDK from the outdated 6.6.2. We've since upgraded to 8.10.0 a week ago, but so far, we've seen no positive changes.**

<details><summary>Some implementation details</summary>
<p>

We've been using the [getToken method](https://firebase.google.com/docs/cloud-messaging/js/client#configure_web_credentials_in_your_app) in a unique way that's not documented:

1. Our service worker's "update" method: 
1.1. Retrieves a token using getToken; 
1.2. Forwards the token to our backend.
2. This "update" method is triggered by: 
2.1. Opening site pages (via postMessage); 
2.2. Worker activation (like after getting a push, installation, etc.).

This procedure has been operational and **problem-free for over 3 years**, but we're concerned that our implementation might not align with [best practices](https://firebase.google.com/docs/cloud-messaging/manage-tokens), potentially contributing to the current issues.

</p>
</details> 

Thanks in advance to everyone who decides to join this discussion. We've been in talks with GCP support for two weeks with no useful assistance, so your suggestions and insights could be incredibly valuable. We appreciate your participation and are ready to answer any questions. 🙏 🙏🙏

![image (1)](https://github.com/firebase/firebase-js-sdk/assets/49155262/0867e0f8-1731-41a5-959d-091af6432387)

**Comment by ramilyayus** [comment, marked as answer]
The issue has been resolved through Google support, no one else should encounter this problem anymore. Despite the fact that version 6.6.2 is still operational, I urge everyone to update to the latest version as soon as possible.

**Reply by AmetVik** [reply]
We have a flutter web project that's precisely facing the same issue and there is no way for me to debug this. Has anyone managed to find the root cause?
